% Template for PLoS
% Version 3.5 March 2018
%
% % % % % % % % % % % % % % % % % % % % % %
%
% -- IMPORTANT NOTE
%
% This template contains comments intended
% to minimize problems and delays during our production
% process. Please follow the template instructions
% whenever possible.
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% Once your paper is accepted for publication,
% PLEASE REMOVE ALL TRACKED CHANGES in this file
% and leave only the final text of your manuscript.
% PLOS recommends the use of latexdiff to track changes during review, as this will help to maintain a clean tex file.
% Visit https://www.ctan.org/pkg/latexdiff?lang=en for info or contact us at latex@plos.org.
%
%
% There are no restrictions on package use within the LaTeX files except that
% no packages listed in the template may be deleted.
%
% Please do not include colors or graphics in the text.
%
% The manuscript LaTeX source should be contained within a single file (do not use \input, \externaldocument, or similar commands).
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% -- FIGURES AND TABLES
%
% Please include tables/figure captions directly after the paragraph where they are first cited in the text.
%
% DO NOT INCLUDE GRAPHICS IN YOUR MANUSCRIPT
% - Figures should be uploaded separately from your manuscript file.
% - Figures generated using LaTeX should be extracted and removed from the PDF before submission.
% - Figures containing multiple panels/subfigures must be combined into one image file before submission.
% For figure citations, please use "Fig" instead of "Figure".
% See http://journals.plos.org/plosone/s/figures for PLOS figure guidelines.
%
% Tables should be cell-based and may not contain:
% - spacing/line breaks within cells to alter layout or alignment
% - do not nest tabular environments (no tabular environments within tabular environments)
% - no graphics or colored text (cell background color/shading OK)
% See http://journals.plos.org/plosone/s/tables for table guidelines.
%
% For tables that exceed the width of the text column, use the adjustwidth environment as illustrated in the example table in text below.
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% -- EQUATIONS, MATH SYMBOLS, SUBSCRIPTS, AND SUPERSCRIPTS
%
% IMPORTANT
% Below are a few tips to help format your equations and other special characters according to our specifications. For more tips to help reduce the possibility of formatting errors during conversion, please see our LaTeX guidelines at http://journals.plos.org/plosone/s/latex
%
% For inline equations, please be sure to include all portions of an equation in the math environment.  For example, x$^2$ is incorrect; this should be formatted as $x^2$ (or $\mathrm{x}^2$ if the romanized font is desired).
%
% Do not include text that is not math in the math environment. For example, CO2 should be written as CO\textsubscript{2} instead of CO$_2$.
%
% Please add line breaks to long display equations when possible in order to fit size of the column.
%
% For inline equations, please do not include punctuation (commas, etc) within the math environment unless this is part of the equation.
%
% When adding superscript or subscripts outside of brackets/braces, please group using {}.  For example, change "[U(D,E,\gamma)]^2" to "{[U(D,E,\gamma)]}^2".
%
% Do not use \cal for caligraphic font.  Instead, use \mathcal{}
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% Please contact latex@plos.org with any questions.
%
% % % % % % % % % % % % % % % % % % % % % % % %

\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}


% Remove comment for double spacing
%\usepackage{setspace}
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother



% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
%\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
%\setlength{\headheight}{27.023pt}
%\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\today}

%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}

%% END MACROS SECTION

\usepackage{latexsym, mathrsfs}
\usepackage{slantsc}
\usepackage{xspace}
\usepackage[titletoc,title]{appendix}
\usepackage{amsthm,amssymb}
\usepackage{subfigure} %{subfig} {subcaption}
\usepackage{multirow}
\usepackage[normalem]{ulem}
\usepackage{url}
\usepackage{multibib}
\newcites{app}{Supplementary References}
%\usepackage[usenames,dvipsnames]{color}
%\usepackage[linesnumbered,ruled,procnumbered]{algorithm2e}
%\usepackage{paralist}
%\usepackage{url}
%\usepackage{graphicx}
%\usepackage{floatflt}
%\usepackage{times}
%\usepackage{balance}
%\usepackage{spverbatim}
%\usepackage{hyperref}

%%%%%%% adjust spacing
\makeatletter
\renewcommand\section{\@startsection{section}{1}{\z@}%
                       {-8\p@ \@plus -4\p@ \@minus -4\p@}%
                       {6\p@ \@plus 4\p@ \@minus 4\p@}%
                       {\normalfont\large\bfseries\boldmath
                        \rightskip=\z@ \@plus 8em\pretolerance=10000 }}
\renewcommand\subsection{\@startsection{subsection}{2}{\z@}%
                       {-8\p@ \@plus -4\p@ \@minus -4\p@}%
                       {6\p@ \@plus 4\p@ \@minus 4\p@}%
                       {\normalfont\normalsize\bfseries\boldmath
                        \rightskip=\z@ \@plus 8em\pretolerance=10000 }}
\renewcommand\subsubsection{\@startsection{subsubsection}{3}{\z@}%
                       {-4\p@ \@plus -4\p@ \@minus -4\p@}%
                       {-1.5em \@plus -0.22em \@minus -0.1em}%
                       {\normalfont\normalsize\bfseries\boldmath}}
\makeatother
\newcommand{\squishlist}{\begin{list}{$\bullet$}{\leftmargin=1.2em \topsep 1pt \itemsep -1pt \parsep 1pt}}
\newcommand{\squishend}{\end{list}}


%%%%%%%%%%% special formatting
\newcommand{\sampar}[1]{\vspace{3pt}\noindent{\bf #1}}
\newcommand{\topic}[1]{\vspace{1pt} \noindent \underline{\bf #1}}
\newcommand{\stitle}[1]{\vspace{0.5em}\noindent\textbf{#1}}


%%%%%%%%%%% Theorem templates
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}{Lemma}
\newtheorem{conjecture}{Conjecture}
\newtheorem{example}{Example}
\newtheorem{formulation}{Problem}
\newtheorem{definition}{Definition}
\newtheorem{observation}[theorem]{\textbf{Observation}}
\newtheorem{claim}[lemma]{\textbf{Claim}}
\newtheorem{problem}{\textbf{Problem}}
%\theoremstyle{slplain}


%%%%%%%%%%%%% special math notation
\newcommand{\mm}{{\mathcal{M}}\xspace}
\newcommand{\ff}{{\mathcal{F}}\xspace}
\newcommand{\oo}{{\mathcal{O}}\xspace}
\newcommand{\sss}{{\mathcal{S}}\xspace}
\newcommand{\cc}{{\mathcal{C}}\xspace}
\newcommand{\lhat}{\mathit{L}\xspace}


%%%%%%%%%%%%%%% special words
\newcommand{\topk}{top-k\xspace}
\newcommand{\topthree}{top-3 }
\newcommand{\toptwenty}{top-20 }
\newcommand{\tophundred}{top-100 }
\newcommand{\topthousand}{top-1000 }
%\newcommand{\topthree}{{\sc T{\small OP-3}}\xspace}
\newcommand{\trans}{{\sc T{\small RANS\-FORMATION}}\xspace}
\newcommand{\genviz}{{\sc G{\small ENVISAGE}}\xspace}
\newcommand{\earlyT}{{\sc E{\small ARLY}S{\small TOP}}\xspace}
\newcommand{\sampling}{{\sc S{\small AMPLING}}\xspace} %\hspace{.2mm}
\newcommand{\traversal}{{\sc T{\small RAVERSAL}}\xspace}
\newcommand{\baseline}{{\sc Baseline}\xspace}
\newcommand{\early}{{\sc Early}\xspace}
\newcommand{\earlyOrder}{{\sc EarlyOrdering}\xspace}
\newcommand{\samp}{{\sc SampOnly}\xspace}
\newcommand{\sampOpt}{{\sc SampOpt}\xspace}
\newcommand{\horiz}{{\sc HorizSampOpt}\xspace}
\newcommand{\vertic}{{\sc VertSampOpt}\xspace}
\newcommand{\msig}{{\sc MSigDB}\xspace}
\newcommand{\lincs}{{\sc LINCS}\xspace}


%%%%%%%%%%%%% editing tools
\newcommand{\sinha}[1]{\noindent{\textcolor{blue}{[Saurabh: #1]}}}
\newcommand{\agp}[1]{\noindent{\textcolor{red}{[Aditya: #1]}}}
\newcommand{\agpins}[1]{\noindent{\textcolor{red}{#1}}}
\newcommand{\cb}[1]{\noindent{\textcolor{cyan}{[Charles: #1]}}}
\newcommand{\silu}[1]{\noindent{\textcolor{magenta}{[Silu: #1]}}}
\newcommand{\xsinha}[1]{\noindent{\textcolor{green}{[X SS: #1]}}}
\newcommand{\xagp}[1]{\noindent{\textcolor{green}{[X AP: #1]}}}
\newcommand{\xcb}[1]{\noindent{\textcolor{green}{[X CB: #1]}}}
\newcommand{\xsilu}[1]{\noindent{\textcolor{green}{[X SH: #1]}}}
\newcommand{\todo}[1]{\noindent{\textcolor{orange}{\textbf{[TODO]: #1}}}}
\newcommand{\needcite}[1]{\noindent{\textcolor{blue}{\textbf{[CITE] #1}}}}
\newcommand{\expn}[1]{\noindent{\textcolor{blue}{[Expand?]: #1}}}
\newcommand{\trim}[1]{\noindent{\textcolor{magenta}{[Reduce?]: #1}}}
\newcommand{\papertext}[1]{\noindent{\textcolor{blue}{PAPER: #1}}}
\newcommand{\cut}[1]{}
\newcommand{\eat}[1]{}
\newcommand{\tr}[1]{}
\newcommand{\tleftout}[1]{}
\newcommand{\new}[1]{\noindent{\textcolor{red}{#1}}}


\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{G{\large ENVISAGE}: Rapid Identification of Discriminative Feature Pairs for Genomic Analysis} % Please use "sentence case" for title and headings (capitalize only the first word in a title (or heading), the first word in a subtitle (or subheading), and any proper nouns).
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Silu Huang\textsuperscript{1\ddag},
Charles Blatti\textsuperscript{2\ddag},
Saurabh Sinha\textsuperscript{1,2\dag},
Aditya Parameswaran\textsuperscript{1,2,3\dag},
%with the Lorem Ipsum Consortium\textsuperscript{\textpilcrow}
\\
\bigskip
\textbf{1} Department of Computer Science, University of Illinois Urbana-Champaign, Urbana, IL, 61801, USA
\\
\textbf{2} Institute of Genomic Biology, University of Illinois Urbana-Champaign, Urbana, IL, 61801, USA
\\
\textbf{3} School of Information and Department of Electrical Engineering and Computer Sciences, UC Berkeley, Berkeley CA, 94704, USA
\bigskip

% Insert additional author notes using the symbols described below. Insert symbol callouts after author names as necessary.
%
% Remove or comment out the author notes below if they aren't used.
%
% Primary Equal Contribution Note

\ddag These authors contributed equally to this work.
\\
\dag To whom correspondence should be addressed.
\\
* sinhas@illinois.edu, adityagp@berkeley.edu
% \Yinyang These authors contributed equally to this work.

% % Additional Equal Contribution Note
% % Also use this double-dagger symbol for special authorship notes, such as senior authorship.
% \ddag These authors also contributed equally to this work.

% % Current address notes
% \textcurrency Current Address: Dept/Program/Center, Institution Name, City, State, Country % change symbol to "\textcurrency a" if more than one current address note
% % \textcurrency b Insert second current address
% % \textcurrency c Insert third current address

% % Deceased author note
% \dag Deceased

% % Group/Consortium Author Note
% \textpilcrow Membership list can be found in the Acknowledgments section.

% Use the asterisk to denote corresponding authorship and provide email address in note below.


\end{flushleft}
% Please keep the abstract below 300 words
\section*{Abstract}
\textbf{Motivation.} \new{A common but critical task in genomic data analysis is 
finding features (often genes) that {\em separate} and potentially explain the 
difference between two classes of biological objects (e.g. healthy versus 
diseased patients).  As lower-cost, high-throughput experimental methods greatly 
increase the number of samples that are assayed as objects for analysis, 
computational methods are needed to quickly provide insights into 
high-dimensional datasets with tens of thousands of objects and features.}\\
\noindent
\textbf{Results.} We develop an \new{interactive exploration} tool called 
\genviz that \new{specializes in rapidly discovering the most discriminative 
feature pairs in a dataset} and \new{displaying} the corresponding 
visualizations. Since quickly finding top feature pairs is computationally 
challenging, especially when the numbers of objects and features are large, we 
propose a suite of optimizations to make \genviz more responsive and demonstrate 
that our optimizations lead to a \new{{\em 400X}} speedup over competitive 
baselines for multiple biological data sets. With this speedup, \genviz enables 
the exploration of more \new{large-scale} datasets \new{and alternate} 
hypotheses in an interactive fashion. We apply \genviz to uncover pairs of genes 
whose transcriptomic responses \new{\sout{more}} significantly discriminate 
treatments of \new{\sout{different} several} chemotherapy drugs.\\

\noindent \textbf{Availability.} Free webserver at 
\url{http://genvisage.knoweng.org:443/} with source code at 
\url{https://github.com/KnowEnG/Genvisage} \\

\linenumbers

\input{body_short.tex}

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step.
% Author Summary not valid for PLOS ONE submissions.
% \section*{Author summary}
% Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur eget porta erat. Morbi consectetur est vel gravida pretium. Suspendisse ut dui eu ante cursus gravida non sed sem. Nullam sapien tellus, commodo id velit id, eleifend volutpat quam. Phasellus mauris velit, dapibus finibus elementum vel, pulvinar non tellus. Nunc pellentesque pretium diam, quis maximus dolor faucibus id. Nunc convallis sodales ante, ut ullamcorper est egestas vitae. Nam sit amet enim ultrices, ultrices elit pulvinar, volutpat risus.



% % Use "Eq" instead of "Equation" for equation citations.
% \section*{Introduction}
% Lorem ipsum dolor sit~\cite{bib1} amet, consectetur adipiscing elit. Curabitur eget porta erat. Morbi consectetur est vel gravida pretium. Suspendisse ut dui eu ante cursus gravida non sed sem. Nullam Eq~(\ref{eq:schemeP}) sapien tellus, commodo id velit id, eleifend volutpat quam. Phasellus mauris velit, dapibus finibus elementum vel, pulvinar non tellus. Nunc pellentesque pretium diam, quis maximus dolor faucibus id.~\cite{bib2} Nunc convallis sodales ante, ut ullamcorper est egestas vitae. Nam sit amet enim ultrices, ultrices elit pulvinar, volutpat risus.

% \begin{eqnarray}
% \label{eq:schemeP}
% 	\mathrm{P_Y} = \underbrace{H(Y_n) - H(Y_n|\mathbf{V}^{Y}_{n})}_{S_Y} + \underbrace{H(Y_n|\mathbf{V}^{Y}_{n})- H(Y_n|\mathbf{V}^{X,Y}_{n})}_{T_{X\rightarrow Y}},
% \end{eqnarray}

% \section*{Materials and methods}
% \subsection*{Etiam eget sapien nibh}

% % For figure citations, please use "Fig" instead of "Figure".
% Nulla mi mi, Fig~\ref{fig1} venenatis sed ipsum varius, volutpat euismod diam. Proin rutrum vel massa non gravida. Quisque tempor sem et dignissim rutrum. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi at justo vitae nulla elementum commodo eu id massa. In vitae diam ac augue semper tincidunt eu ut eros. Fusce fringilla erat porttitor lectus cursus, \nameref{S1_Video} vel sagittis arcu lobortis. Aliquam in enim semper, aliquam massa id, cursus neque. Praesent faucibus semper libero.

% % Place figure captions after the first paragraph in which they are cited.
% \begin{figure}[!h]
% \caption{{\bf Bold the figure title.}
% Figure caption text here, please use this space for the figure panel descriptions instead of using subfigure commands. A: Lorem ipsum dolor sit amet. B: Consectetur adipiscing elit.}
% \label{fig1}
% \end{figure}

% % Results and Discussion can be combined.
% \section*{Results}
% Nulla mi mi, venenatis sed ipsum varius, Table~\ref{table1} volutpat euismod diam. Proin rutrum vel massa non gravida. Quisque tempor sem et dignissim rutrum. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi at justo vitae nulla elementum commodo eu id massa. In vitae diam ac augue semper tincidunt eu ut eros. Fusce fringilla erat porttitor lectus cursus, vel sagittis arcu lobortis. Aliquam in enim semper, aliquam massa id, cursus neque. Praesent faucibus semper libero.

% % Place tables after the first paragraph in which they are cited.
% \begin{table}[!ht]
% \begin{adjustwidth}{-2.25in}{0in} % Comment out/remove adjustwidth environment if table fits in text column.
% \centering
% \caption{
% {\bf Table caption Nulla mi mi, venenatis sed ipsum varius, volutpat euismod diam.}}
% \begin{tabular}{|l+l|l|l|l|l|l|l|}
% \hline
% \multicolumn{4}{|l|}{\bf Heading1} & \multicolumn{4}{|l|}{\bf Heading2}\\ \thickhline
% $cell1 row1$ & cell2 row 1 & cell3 row 1 & cell4 row 1 & cell5 row 1 & cell6 row 1 & cell7 row 1 & cell8 row 1\\ \hline
% $cell1 row2$ & cell2 row 2 & cell3 row 2 & cell4 row 2 & cell5 row 2 & cell6 row 2 & cell7 row 2 & cell8 row 2\\ \hline
% $cell1 row3$ & cell2 row 3 & cell3 row 3 & cell4 row 3 & cell5 row 3 & cell6 row 3 & cell7 row 3 & cell8 row 3\\ \hline
% \end{tabular}
% \begin{flushleft} Table notes Phasellus venenatis, tortor nec vestibulum mattis, massa tortor interdum felis, nec pellentesque metus tortor nec nisl. Ut ornare mauris tellus, vel dapibus arcu suscipit sed.
% \end{flushleft}
% \label{table1}
% \end{adjustwidth}
% \end{table}


% %PLOS does not support heading levels beyond the 3rd (no 4th level headings).
% \subsection*{\lorem\ and \ipsum\ nunc blandit a tortor}
% \subsubsection*{3rd level heading}
% Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque. Quisque augue sem, tincidunt sit amet feugiat eget, ullamcorper sed velit. Sed non aliquet felis. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris commodo justo ac dui pretium imperdiet. Sed suscipit iaculis mi at feugiat.

% \begin{enumerate}
% 	\item{react}
% 	\item{diffuse free particles}
% 	\item{increment time by dt and go to 1}
% \end{enumerate}


% \subsection*{Sed ac quam id nisi malesuada congue}

% Nulla mi mi, venenatis sed ipsum varius, volutpat euismod diam. Proin rutrum vel massa non gravida. Quisque tempor sem et dignissim rutrum. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi at justo vitae nulla elementum commodo eu id massa. In vitae diam ac augue semper tincidunt eu ut eros. Fusce fringilla erat porttitor lectus cursus, vel sagittis arcu lobortis. Aliquam in enim semper, aliquam massa id, cursus neque. Praesent faucibus semper libero.

% \begin{itemize}
% 	\item First bulleted item.
% 	\item Second bulleted item.
% 	\item Third bulleted item.
% \end{itemize}

% \section*{Discussion}
% Nulla mi mi, venenatis sed ipsum varius, Table~\ref{table1} volutpat euismod diam. Proin rutrum vel massa non gravida. Quisque tempor sem et dignissim rutrum. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi at justo vitae nulla elementum commodo eu id massa. In vitae diam ac augue semper tincidunt eu ut eros. Fusce fringilla erat porttitor lectus cursus, vel sagittis arcu lobortis. Aliquam in enim semper, aliquam massa id, cursus neque. Praesent faucibus semper libero~\cite{bib3}.

% \section*{Conclusion}

% CO\textsubscript{2} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque. Quisque augue sem, tincidunt sit amet feugiat eget, ullamcorper sed velit.

% Sed non aliquet felis. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris commodo justo ac dui pretium imperdiet. Sed suscipit iaculis mi at feugiat. Ut neque ipsum, luctus id lacus ut, laoreet scelerisque urna. Phasellus venenatis, tortor nec vestibulum mattis, massa tortor interdum felis, nec pellentesque metus tortor nec nisl. Ut ornare mauris tellus, vel dapibus arcu suscipit sed. Nam condimentum sem eget mollis euismod. Nullam dui urna, gravida venenatis dui et, tincidunt sodales ex. Nunc est dui, sodales sed mauris nec, auctor sagittis leo. Aliquam tincidunt, ex in facilisis elementum, libero lectus luctus est, non vulputate nisl augue at dolor. For more information, see \nameref{S1_Appendix}.

% \section*{Supporting information}

% % Include only the SI item label in the paragraph heading. Use the \nameref{label} command to cite SI items in the text.
% \paragraph*{S1 Fig.}
% \label{S1_Fig}
% {\bf Bold the title sentence.} Add descriptive text after the title of the item (optional).

% \paragraph*{S2 Fig.}
% \label{S2_Fig}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 File.}
% \label{S1_File}
% {\bf Lorem ipsum.}  Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Video.}
% \label{S1_Video}
% {\bf Lorem ipsum.}  Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Appendix.}
% \label{S1_Appendix}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Table.}
% \label{S1_Table}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \section*{Acknowledgments}
% Cras egestas velit mauris, eu mollis turpis pellentesque sit amet. Interdum et malesuada fames ac ante ipsum primis in faucibus. Nam id pretium nisi. Sed ac quam id nisi malesuada congue. Sed interdum aliquet augue, at pellentesque quam rhoncus vitae.
%\input{appendix.tex}

\nolinenumbers

% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% or
%
% Compile your BiBTeX database using our plos2015.bst
% style file and paste the contents of your .bbl file
% here. See http://journals.plos.org/plosone/s/latex for
% step-by-step instructions.
%


% \bibliographystyle{plain}
% \bibliography{ref}

\begin{thebibliography}{10}

\bibitem{tang2017gepia}
Tang Z, et~al.
\newblock GEPIA: a web server for cancer and normal gene expression profiling
  and interactive analyses.
\newblock Nucleic acids research. 2017;45(W1):W98--W102.

\bibitem{lamb2006connectivity}
Lamb J, et~al.
\newblock The Connectivity Map: using gene-expression signatures to connect
  small molecules, genes, and disease.
\newblock science. 2006;313(5795):1929--1935.

\bibitem{liberzon2015molecular}
Liberzon A, et~al.
\newblock The molecular signatures database hallmark gene set collection.
\newblock Cell systems. 2015;1(6):417--425.

\bibitem{dudoit2002statistical}
Dudoit S, et~al.
\newblock Statistical methods for identifying differentially expressed genes in
  replicated cDNA microarray experiments.
\newblock Statistica sinica. 2002; p. 111--139.

\bibitem{lai2006comparison}
Lai C, et~al.
\newblock A comparison of univariate and multivariate gene selection techniques
  for classification of cancer datasets.
\newblock BMC bioinformatics. 2006;7(1):235.

\bibitem{peng2010regularized}
Peng J, et~al.
\newblock Regularized multivariate regression for identifying master predictors
  with application to integrative genomics study of breast cancer.
\newblock The annals of applied statistics. 2010;4(1):53.

\bibitem{breiman2001random}
Breiman L.
\newblock Random forests.
\newblock Machine learning. 2001;45(1):5--32.

\bibitem{hanczar2007feature}
Hanczar B, Zucker JD, Henegar C, Saitta L.
\newblock Feature construction from synergic pairs to improve microarray-based
  classification.
\newblock Bioinformatics. 2007;23(21):2866--2872.

\bibitem{geman2004classifying}
Geman D, d'Avignon C, Naiman DQ, Winslow RL.
\newblock Classifying gene expression profiles from pairwise mRNA comparisons.
\newblock Statistical applications in genetics and molecular biology.
  2004;3(1):1--19.

\bibitem{shi2011top}
Shi P, Ray S, Zhu Q, Kon MA.
\newblock Top scoring pairs for feature selection in machine learning and
  applications to cancer outcome prediction.
\newblock BMC bioinformatics. 2011;12(1):375.

\bibitem{shen2017identification}
Shen R, Luo L, Jiang H.
\newblock Identification of gene pairs through penalized regression subject to
  constraints.
\newblock BMC bioinformatics. 2017;18(1):466.

\bibitem{sinha2008systematic}
Sinha S, Adler AS, Field Y, Chang HY, Segal E.
\newblock Systematic functional characterization of cis-regulatory motifs in
  human core promoters.
\newblock Genome research. 2008;18(3):477--488.

\bibitem{watkinson2008identification}
Watkinson J, Wang X, Zheng T, Anastassiou D.
\newblock Identification of gene interactions associated with disease from gene
  expression data using synergy networks.
\newblock BMC systems biology. 2008;2(1):10.

\bibitem{basu2018iterative}
Basu S, Kumbier K, Brown JB, Yu B.
\newblock Iterative random forests to discover predictive and stable high-order
  interactions.
\newblock Proceedings of the National Academy of Sciences.
  2018;115(8):1943--1948.

\bibitem{shah2014random}
Shah RD, Meinshausen N.
\newblock Random intersection trees.
\newblock The Journal of Machine Learning Research. 2014;15(1):629--654.

\bibitem{schwarz2010safari}
Schwarz DF, K{\"o}nig IR, Ziegler A.
\newblock On safari to Random Jungle: a fast implementation of Random Forests
  for high-dimensional data.
\newblock Bioinformatics. 2010;26(14):1752--1758.

\bibitem{unger2010linear}
Unger G, Chor B.
\newblock Linear separability of gene expression data sets.
\newblock IEEE/ACM Transactions on Computational Biology and Bioinformatics
  (TCBB). 2010;7(2):375--381.

\bibitem{sinha2003discriminative}
Sinha S.
\newblock Discriminative motifs.
\newblock Journal of Computational Biology. 2003;10(3-4):599--615.

\bibitem{medin1981linear}
Medin D, Schwanenflugel P.
\newblock Linear separability in classification learning.
\newblock Journal of Experimental Psychology: Human Learning and Memory.
  1981;7(5):355.

\bibitem{vapnik1998statistical}
Vapnik V, Vapnik V.
\newblock Statistical learning theory.
\newblock Wiley New York; 1998.

\bibitem{rocchio1971relevance}
Rocchio J.
\newblock Relevance feedback in information retrieval.
\newblock The SMART retrieval system: experiments in automatic document
  processing. 1971; p. 313--323.

\bibitem{hoeffding1963probability}
Hoeffding W.
\newblock Probability inequalities for sums of bounded random variables.
\newblock Journal of the American statistical association. 1963;58(301):13--30.

\bibitem{blatti2016characterizing}
Blatti C, Sinha S.
\newblock Characterizing gene sets using discriminative random walks with
  restart on heterogeneous biological networks.
\newblock Bioinformatics. 2016;32(14).

\bibitem{subramanian2005gene}
Subramanian A, et~al.
\newblock Gene set enrichment analysis: a knowledge-based approach for
  interpreting genome-wide expression profiles.
\newblock Proceedings of the National Academy of Sciences.
  2005;102(43):15545--15550.

\bibitem{subramanian2017next}
Subramanian A, et~al.
\newblock A next generation connectivity map: L1000 platform and the first
  1,000,000 profiles.
\newblock Cell. 2017;171(6):1437--1452.

\bibitem{grondin2018accessing}
Grondin C, et~al.
\newblock Accessing an Expanded Exposure Science Module at the Comparative
  Toxicogenomics Database.
\newblock Environmental health perspectives. 2018;126(1).

\bibitem{chan2012serine}
Chan H, et~al.
\newblock Serine protease PRSS23 is upregulated by estrogen receptor $\alpha$
  and associated with proliferation of breast cancer cells.
\newblock PLoS One. 2012;7(1):e30397.

\bibitem{moggs2004phenotypic}
Moggs J, et~al.
\newblock Phenotypic anchoring of gene expression changes during
  estrogen-induced uterine growth.
\newblock Environmental health perspectives. 2004;112(16):1589.

\bibitem{spink2009long}
Spink B, et~al.
\newblock Long-term estrogen exposure promotes carcinogen bioactivation,
  induces persistent changes in gene expression, and enhances the
  tumorigenicity of MCF-7 human breast cancer cells.
\newblock Toxicology and applied pharmacology. 2009;240(3).

\bibitem{sengupta2013molecular}
Sengupta S, et~al.
\newblock Molecular mechanism of action of bisphenol and bisphenol A mediated
  by oestrogen receptor alpha in growth and apoptosis of breast cancer cells.
\newblock British journal of pharmacology. 2013;169(1):167--178.

\bibitem{qi2014systematic}
Qi Y, et~al.
\newblock Systematic analysis of time-series gene expression data on tumor
  cell-selective apoptotic responses to HDAC inhibitors.
\newblock Computational and mathematical methods in medicine. 2014;2014.

\bibitem{soldi2013genomic}
Soldi R, et~al.
\newblock A genomic approach to predict synergistic combinations for breast
  cancer treatment.
\newblock The pharmacogenomics journal. 2013;13(1):94.

\bibitem{zhao2015gene}
Zhao W, et~al.
\newblock Gene expression profiling identifies the novel role of
  immunoproteasome in doxorubicin-induced cardiotoxicity.
\newblock Toxicology. 2015;333:76--88.

\bibitem{gorka2007nrcam}
Gorka B, et~al.
\newblock NrCAM, a neuronal system cell-adhesion molecule, is induced in
  papillary thyroid carcinomas.
\newblock British journal of cancer. 2007;97(4):531.

\bibitem{gene2014gene}
Consortium GO.
\newblock Gene ontology consortium: going forward.
\newblock Nucleic acids research. 2014;43(D1):D1049--D1056.

\bibitem{finn2015pfam}
Finn R, et~al.
\newblock The Pfam protein families database: towards a more sustainable
  future.
\newblock Nucleic acids research. 2015;44(D1):D279--D285.

\bibitem{fabregat2017reactome}
Fabregat A, et~al.
\newblock The reactome pathway knowledgebase.
\newblock Nucleic acids research. 2017;46(D1):D649--D655.

\bibitem{kanehisa2016kegg}
Kanehisa M, et~al.
\newblock KEGG: new perspectives on genomes, pathways, diseases and drugs.
\newblock Nucleic acids research. 2016;45(D1):D353--D361.

\bibitem{altschul1990basic}
Altschul S, et~al.
\newblock Basic local alignment search tool.
\newblock Journal of molecular biology. 1990;215(3):403--410.

\bibitem{szklarczyk2014string}
Szklarczyk D, et~al.
\newblock STRING v10: protein--protein interaction networks, integrated over
  the tree of life.
\newblock Nucleic acids research. 2014;43(D1):D447--D452.

\bibitem{croft2013reactome}
Croft D, et~al.
\newblock The Reactome pathway knowledgebase.
\newblock Nucleic acids research. 2013;42(D1):D472--D477.

\bibitem{cerami2010pathway}
Cerami E, et~al.
\newblock Pathway Commons, a web resource for biological pathway data.
\newblock Nucleic acids research. 2010;39(suppl\_1):D685--D690.

\bibitem{lee2011prioritizing}
Lee I, et~al.
\newblock Prioritizing candidate disease genes by network-based boosting of
  genome-wide association data.
\newblock Genome research. 2011;21(7):1109--1121.

\bibitem{chatr2017biogrid}
Chatr-Aryamontri A, et~al.
\newblock The BioGRID interaction database: 2017 update.
\newblock Nucleic acids research. 2017;45(D1):D369--D379.

\bibitem{orchard2013mintact}
Orchard S, et~al.
\newblock The MIntAct project-IntAct as a common curation platform for 11
  molecular interaction databases.
\newblock Nucleic acids research. 2013;42(D1):D358--D363.

\bibitem{salwinski2004database}
Salwinski L, et~al.
\newblock The database of interacting proteins: 2004 update.
\newblock Nucleic acids research. 2004;32(suppl\_1):D449--D451.

\end{thebibliography}



\clearpage
\begin{appendices}
\input{appendix.tex}
\end{appendices}

\end{document}

