\documentclass{bioinfo}

\usepackage{latexsym, mathrsfs}
\usepackage{slantsc}
%\usepackage[usenames,dvipsnames]{color}
%\usepackage[linesnumbered,ruled,procnumbered]{algorithm2e}
%\usepackage{paralist}
\usepackage{xspace}
%\usepackage{url}
%\usepackage{graphicx}
%\usepackage{floatflt}
%\usepackage{times}
%\usepackage{balance}
%\usepackage{spverbatim}
%\usepackage{hyperref}
\usepackage{amsthm,amssymb}
\usepackage{subfigure} %{subfig} {subcaption}
\usepackage{multirow}
\usepackage[normalem]{ulem}
\usepackage{url}


\newcommand{\topic}[1]{\vspace{1pt} \noindent \underline{\bf #1}}
%\theoremstyle{slplain}
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}{Lemma}
\newtheorem{conjecture}{Conjecture}
\newtheorem{example}{Example}
\newtheorem{formulation}{Problem}
\newtheorem{definition}{Definition}
\newtheorem{observation}[theorem]{\textbf{Observation}}
\newtheorem{claim}[lemma]{\textbf{Claim}}
\newtheorem{problem}{\textbf{Problem}}
\newcommand{\stitle}[1]{\vspace{0.5em}\noindent\textbf{#1}}
\newcommand{\mm}{{\mathcal{M}}\xspace}
\newcommand{\ff}{{\mathcal{F}}\xspace}
\newcommand{\oo}{{\mathcal{O}}\xspace}
\newcommand{\sss}{{\mathcal{S}}\xspace}
\newcommand{\cc}{{\mathcal{C}}\xspace}
\newcommand{\lhat}{\mathit{L}\xspace}
\newcommand{\topk}{top-k }
\newcommand{\topthree}{top-3 }
\newcommand{\toptwenty}{top-20 }
\newcommand{\tophundred}{top-100 }
\newcommand{\topthousand}{top-1000 }
%\newcommand{\topthree}{{\sc T{\small OP-3}}\xspace}
\newcommand{\trans}{{\sc T{\small RANS\-FORMATION}}\xspace}
\newcommand{\genviz}{{\sc G{\small ENVISAGE}}\xspace}
\newcommand{\earlyT}{{\sc E{\small ARLY}S{\small TOP}}\xspace}
\newcommand{\sampling}{{\sc S{\small AMPLING}}\xspace} %\hspace{.2mm}
\newcommand{\traversal}{{\sc T{\small RAVERSAL}}\xspace}
\newcommand{\baseline}{{\sc Baseline}\xspace}
\newcommand{\early}{{\sc Early}\xspace}
\newcommand{\earlyOrder}{{\sc EarlyOrdering}\xspace}
\newcommand{\samp}{{\sc SampOnly}\xspace}
\newcommand{\sampOpt}{{\sc SampOpt}\xspace}
\newcommand{\horiz}{{\sc HorizSampOpt}\xspace}
\newcommand{\vertic}{{\sc VertSampOpt}\xspace}
\newcommand{\msig}{{\sc MSigDB}\xspace}
\newcommand{\lincs}{{\sc LINCS}\xspace}
\newcommand{\sinha}[1]{\noindent{\textcolor{blue}{[Saurabh: #1]}}}
\newcommand{\agp}[1]{\noindent{\textcolor{red}{[Aditya: #1]}}}
\newcommand{\agpins}[1]{\noindent{\textcolor{red}{#1}}}
\newcommand{\cb}[1]{\noindent{\textcolor{cyan}{[Charles: #1]}}}
\newcommand{\silu}[1]{\noindent{\textcolor{magenta}{[Silu: #1]}}}
\newcommand{\xsinha}[1]{\noindent{\textcolor{green}{[X SS: #1]}}}
\newcommand{\xagp}[1]{\noindent{\textcolor{green}{[X AP: #1]}}}
\newcommand{\xcb}[1]{\noindent{\textcolor{green}{[X CB: #1]}}}
\newcommand{\xsilu}[1]{\noindent{\textcolor{green}{[X SH: #1]}}}
\newcommand{\todo}[1]{\noindent{\textcolor{orange}{\textbf{[TODO]: #1}}}}
\newcommand{\needcite}[1]{\noindent{\textcolor{blue}{\textbf{[CITE] #1}}}}
\newcommand{\expn}[1]{\noindent{\textcolor{blue}{[Expand?]: #1}}}
\newcommand{\cut}[1]{}
\newcommand{\trim}[1]{\noindent{\textcolor{magenta}{[Reduce?]: #1}}}
\newcommand{\eat}[1]{}
\newcommand{\papertext}[1]{\noindent{\textcolor{blue}{PAPER: #1}}}
%\newcommand{\tr}[1]{\noindent{\textcolor{red}{#1}}}
\newcommand{\tr}[1]{}
%\newcommand{\tleftout}[1]{\noindent{\textcolor{blue}{#1}}}
\newcommand{\tleftout}[1]{}
\newcommand{\squishlist}{\begin{list}{$\bullet$}{\leftmargin=1.2em \topsep 1pt \itemsep -1pt \parsep 1pt}}
\newcommand{\squishend}{\end{list}}
\newcommand{\sampar}[1]{\vspace{3pt}\noindent{\bf #1}}


\copyrightyear{2018} \pubyear{2018}
\access{} %Advance Access Publication Date: Day Month Year
\appnotes{Original Paper} %Manuscript Category

\begin{document}
\firstpage{1}

\subtitle{} %Subject Section

\title[Genvisage]{G{\large ENVISAGE}: Rapid Identification of Discriminative Feature Pairs for Genomic Analysis}

\author[Huang \textit{et~al}.]{Silu Huang\,$^{\text{\sfb 1}}$, Charles Blatti\,$^{\text{\sfb 2}}$, Aditya Parameswaran\,$^{\text{\sfb 1}}$, and Saurabh Sinha\,$^{\text{\sfb 1,2}*}$}

\address{$^{\text{\sf 1}}$Department of Computer Science, University of Illinois Urbana-Champaign, Urbana, IL, 61801, USA and \\
$^{\text{\sf 2}}$Institute of Genomic Biology, University of Illinois Urbana-Champaign, Urbana, IL, 61801, USA.}

\author[]{} %[Sample \textit{et~al}.]{}

% \authorinfo{xxx}
%        {University of Illinois (UIUC)}
%         {\url{xxx@illinois.edu}}



\address{} %$^{\text{\sf 1}}$Computer Science, University of Illinois, Urbana-Champaign, 61802, US.

\corresp{} %$^\ast$To whom correspondence should be addressed.

\history{} %Received on XXXXX; revised on XXXXX; accepted on XXXXX

\editor{} %Associate Editor: XXXXXXX

%\abstract{
%A common but critical task in biological data analysis is the {\em separability} task: finding features that explain the difference between two different classes of objects with high dimensional feature representations, such as genes described by their functional annotations, or gene signatures described by their transcriptomic profiles. We develop an interactive data exploration tool called \genviz for this task that rapidly identifies discriminative feature pairs and outputs the corresponding visualizations. Since quickly finding top-k feature pairs is computationally challenging, especially when the numbers of objects and features are large, we propose a suite of optimizations to make \genviz more responsive, and demonstrate that our optimizations lead to a 400X speedup over competitive baselines for multiple biological data sets. With this speedup, \genviz enables the exploration of more datasets and alternative hypotheses in an interactive fashion. Finally, we apply \genviz to uncover pairs of genes whose transcriptomic responses more significantly discriminate treatments of a given drug.}
\abstract{\textbf{Motivation:} A common but critical task in biological data analysis is the {\em separability} task: finding features that explain the difference between two different classes of objects with high dimensional feature representations, such as genes described by their functional annotations, or gene signatures described by their transcriptomic profiles.\\
\textbf{Results:} We develop an interactive data exploration tool called \genviz for this task that rapidly identifies discriminative feature pairs and outputs the corresponding visualizations. Since quickly finding top-k feature pairs is computationally challenging, especially when the numbers of objects and features are large, we propose a suite of optimizations to make \genviz more responsive, and demonstrate that our optimizations lead to a 400X speedup over competitive baselines for multiple biological data sets. With this speedup, \genviz enables the exploration of more datasets and alternative hypotheses in an interactive fashion. Finally, we apply \genviz to uncover pairs of genes whose transcriptomic responses more significantly discriminate treatments of a given drug.\\
\textbf{Availability:} Free webserver at \href{http://128.174.244.49:5000}{http://128.174.244.49:5000} \silu{move to permanent/accessible domain} with source code at \href{https://github.com/KnowEnG/Genvisage}{https://github.com/KnowEnG/Genvisage} \\
\textbf{Contact:} \href{sinhas@illinois.edu}{sinhas@illinois.edu}\\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.}

\maketitle

\input{body.tex}

{\scriptsize
\bibliographystyle{abbrv}
\bibliography{ref}
}
\end{document}










